designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+
+The underlying type system for Stan relies on a notion of runtime
+typing of all objects.  The runtime type does *not* include any
+constraints.  The constraints are used for bounds checking for read or
+constructed types and for transforms for parameters.
+
+Sized types are used for block declarations and unsized types are used
+for function arguments.  Local variables are currently sized, but in
+the future will be allowed to be unsized.
+
+
+#### Unsized runtime types
+
+The set of unsized runtime types is the least set of types such that
+
+* *unsized primitive type*: `int` and `real` are unsized types,
+* *unsized vector type*: `vector` and `row_vector` are unsized types,
+* *unsized matrix typen*: , `matrix` is an unsized type,
+* *unsized array type*: `T[]` is an unsized type if `T` is an unsized type, and
+* *unsized function type*: `T0(T1,...,TN)` is an unsized type if `T0`, `T1`, ..., `TN`
+are unsized types.
+
+
+
+#### Covariance and contravariance
+
+If we wanted to get fancy, we could use the polarity of type
+occurrences in function types to extend assignability.  For the base
+case, we have `int` as a subtype of `real` in the sense that we can
+assign an `int` expression to a `real` variable but not vice-versa.
+Covariant typing would extend this as follows.
+
+* `int` is a subtype `real`,
+* `T[]` is a subtype of `U[]` if `T` is a subtype of `U`, and
+* `T0(T1, ..., TN)` is a subtype of `U0(U1, ..., UN)` if
+  `T0` is a subtype of `U0`, `U1` is a subtype of `T1`, ..., `UN` is a
+  subtype of `TN`.
+
+The type `T` in the array type `T[]` is covariant in that it preserves
+subtyping.  The type `T` in `T(U)` is similarly covariant, but `U` is
+contravariant, in that it reverses the subtyping relationship.  As a
+concrete example, consider legal types for `a` (the lvalue) and `b`
+(the rvalue) in `a = b`,
+
+| lvalue type  | legal rvalue types                     |
+|:------------:|:--------------------------------------:|
+| `int(int)`   | `int(int)`, `int(real)`                |
+| `real(real)` | `real(real)`, `int(real)`              |
+| `int(real)`  | `int(real)`                            |
+| `real(int)`  | `real(int)`, `real(real)`, `int(real)` |
+
+
+#### Sized types
+
+The set of sized runtime types is the least set of types such that
+
+* *sized primitive type*: `int` and `real` are sized types,
+* *sized vector type*: `vector[K]` and `row_vector[K]` are sized
+types if `K` is a non-negative integer,
+* *sized matrix type*: `matrix[M, N]` is a sized type if `M` and `N` are
+  non-negative integers,
+* *sized array type*: `T[]` is a sized type if `T` is a sized type, and
+* *sized function type*: `T0(T1,...,TN)` is a sized type if `T0`, `T1`, ..., `TN`
+are sized types.
+
+
+## Lambda Expression Syntax
+
+Syntactically, a lambda expression is represented as a function argument
+list followed by a function body.   For example, in `(real u) { return
+1 / (1 + exp(-u)); }`, the function argument list is `(real u)` and
```

Comment:
Is it worth considering introducing a keyword for lambdas? For example `\` or `fun` or `function`? I'm asking because the parentheses are already used for a lot of different things in Stan, as is, and probably will be used for more if we ever add tuples (unless we stick to structs instead). This is a genuine question. I'm not sure what I think, myself.